### PR TITLE
Planck corner case

### PR DESCRIPTION
**Pull request by kgbudge**
### Background

* The CDI functions to integrate Planck/Rosseland spectrums have some very krufty preconditions for cold cell corner cases.
* These create significant inconveniences for clients of these functions.

### Purpose of Pull Request

* Modify functions to handle T==0 and T << bounds corner cases well.
* FIx a derp in previous PR for extended boundaries policy.
* Fix precondition that lets a group bounds specification with only one bound slip past.

### Status

* Reference:
  * [Pre-Merge Code Review](https://github.com/lanl/Draco/wiki/Style-Guide)
  * [CDash Status](https://rtt.lanl.gov/cdash/index.php?project=Draco&filtercount=1&showfilters=1&field1=buildname&compare1=63&value1=-pr)
* Checks
  * [x] Travis/Appveyor CI checks pass
  * [x] Code coverage does not decrease
  * [x] [Valgrind test passes](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Toss3 checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] [Trinitite checks pass](https://rtt.lanl.gov/CDash/index.php?project=Draco)
  * [x] Code reviewed/approved, sufficient DbC checks, testing, documentation
  * [x] Manual check on trinitite to triage CI test failure (KT)
  * [x] Test downstream impact on other codes (KT)

### CODE REVIEW

**Comment by attom** [comment]
I only see one file committed (a test). Was there a source file that got missed?

**Comment by KineticTheory** [comment]
@kgbudge CI for optimized builds on trinitite are failing.  https://rtt.lanl.gov/cdash3/test/2833950.  Can you look at it?

**Comment by KineticTheory** [comment]
I don't think the ccscs6 valgrind errors are related to this PR.  I'm asking @brryan about these.

**Comment by kgbudge** [comment]
> @kgbudge CI for optimized builds on trinitite are failing. https://rtt.lanl.gov/cdash3/test/2833950. Can you look at it?

It seems to build and run successfully on both KNL and Haswell.

**Comment by KineticTheory** [comment]
> 
> 
> > @kgbudge CI for optimized builds on trinitite are failing. https://rtt.lanl.gov/cdash3/test/2833950. Can you look at it?
> 
> It seems to build and run successfully on both KNL and Haswell.

I'll need to look closer.  The failure is for optimized build with intel on Haswell:

```
Test: failed
     planck extended integral, first form
Test: failed
     planck extended integral, second form
Test: failed
     rosseland extended integral
Test: failed
     planck extended integral, paired
Test: failed
     rosseland extended integral, paired
```

I'll checkout your PR and try it.

File: src/cdi/CDI.cc [hunk 1]
```diff
@@ -211,36 +211,43 @@ void CDI::integrate_Rosseland_Planckian_Spectrum(const size_t groupIndex, const
...
-    Ensure(planck[group] >= 0.0);
-    Ensure(planck[group] <= 1.0);
-  }
-  if (extend) {
-    planck[groups - 1] += 1 - planck_value;
+  if (T > 0.0) {
+    // Initialize the loop:
+    double frequency = bounds[0];
+    double planck_value =
+        extend ? 0.0 : ((frequency < T * 50) ? integrate_planck(frequency / T) : 1.0);
```
Comment (KineticTheory) [inline src/cdi/CDI.cc:232]: `frequency / T` can overflow if `T` is too small.  Can you re-incorporate something akin to this line that you removed?
```c++
  // nu/T must be < numeric_limits<double>::max(), throw assertion in DBC mode if this occurs
  Check(T > bounds.back() * std::numeric_limits<double>::min());
```
One possibility would be changing the conditional `if(T>0.0)` to `if(T>bounds[0]*std::numeric_limits<double>::min())` and adjusting the `else` portion accordingly for `T=epsilon`.

A similar piece of logic will be required below where you call `integrate_planck(frequency / T)`

Comment (KineticTheory) [reply src/cdi/CDI.cc:232]: The magic number `50` should be documented.  I don't understand why this value appears here (and below).

**Review by KineticTheory — CHANGES_REQUESTED** [review]
* I am requesting two small changes (documentation of magic numbers and protection for overflow).
* I will add check boxes to cover the following testing that I will do:
  * Test this PR for downstream impact in my code.
  * Test on trinitite since CI shows a failure.
  *

File: src/cdi/CDI.cc (hunk 1 above)
Comment (kgbudge) [reply src/cdi/CDI.cc:232]: These two go together.  If T is zero, there's a special branch to handle that. If T is nonzero, frequency< T * 50 guarantees the division is well-behaved and cannot overflow. The magic value T*50 is the point on the upper tail where the integral becomes 1 to roundoff. (Actually, the value is around 48; choosing 50 gives just a little more slack if a compiler does something weird and wonderful with its optimizations.)

But, yeah, it seems clear I need to document that.

**Comment by KineticTheory** [comment]
I'll do the extra trinitite testing once it comes back from DST.

**Review by KineticTheory — APPROVED** [review]

**Pull request merged**